### ISSUE TITLE
Error updating single value navigation properties

### ISSUE THREAD

**Issue by PaladinMattt** [jlattimer/CRMRESTBuilder#7]
I have a customer lookup attribute on the task entity (pdc_trainer) which is a lookup of type system user.  I get an error "Bad Request" 
The specific error is: An undeclared property 'pdc_trainer' which only has property annotations in the payload but no property value was found in the payload.

**Comment by jlattimer** [comment]
Any chance you could check Fiddler and post the error detail?

**Comment by PaladinMattt** [comment]
I’m new to using Fiddler, so hopefully this is what you want:
{
  "error":{
    "code":"","message":"An undeclared property 'pdc_trainer' which only has property annotations in the payload but no property value was found in the payload. In OData, only declared navigation properties and declared named streams can be represented as properties without values.","innererror":{
      "message":"An undeclared property 'pdc_trainer' which only has property annotations in the payload but no property value was found in the payload. In OData, only declared navigation properties and declared named streams can be represented as properties without values.","type":"Microsoft.Crm.CrmHttpException","stacktrace":"   at Microsoft.Crm.Extensibility.OData.CrmODataUtilities.ValidateInputParameters(ModelStateDictionary controllerModelState)\r\n   at Microsoft.Crm.Extensibility.OData.EntityController.PatchEntity(String entityName, String key, EdmEntityObject entityDelta)\r\n   at lambda_method(Closure , Object , Object[] )\r\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)\r\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.Execute(Object instance, Object[] arguments)\r\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()"
    }
  }
}


Matt Thompson
In-house Engineer | Paladin Data Corp
paladinpointofsale.com | 1.800.725.2346 ext. 7013

[2014 color horizontal-email signature]

**Comment by jlattimer** [comment]
Look in the metadata located here: https://**org**.crm.dynamics.com/api/data/v**8.2**/$metadata and search for pdc_trainer. You'll probably see it listed as pdc_Trainer_Task - this is what you'll need. 

The code would end up being:
`entity["pdc_Trainer_Task@odata.bind"] = "/systemusers(D23DF1C1-6D52-43F7-968B-94B93CF434CA)";`

I had someone run into this the other day, I'll need to see if I can't account for this when generating code.

**Comment by PaladinMattt** [comment]
Thanks,
It was “pdc_trainer_Task”.  If you figure out how to account for it please let me know.  I was building a wrapper for the web api and ran into the same problem with my code, which is when I noticed yours was doing it too.


Matt Thompson
In-house Engineer | Paladin Data Corp
paladinpointofsale.com | 1.800.725.2346 ext. 7013

[2014 color horizontal-email signature]

**Comment by jamesnovak** [comment]
Ran into the same issue.  In the attribute selector for Create, looks like the correct Schema Name is being used, but in the gen'd code, it's all lower case: 
![untitled](https://cloud.githubusercontent.com/assets/69029/21155968/24822570-c142-11e6-9378-e98bcc5f1a43.png)

Excellent tool!

**Comment by PaladinMattt** [comment]
Jason, I had some time to look into this today and you can query the navigational property in v8.2:
https://yourdomain.crm.dynamics.com/api/data/v8.2/RelationshipDefinitions/Microsoft.Dynamics.CRM.OneToManyRelationshipMetadata?$select=ReferencingEntityNavigationPropertyName&$filter=ReferencingAttribute eq 'theAttributeName' and ReferencingEntity eq 'theReferencingEntityLogicalName'

So in my program, everytime I want to update a lookup attribute, I just make a call to this endpoint and get the navigational property name.

**Comment by jlattimer** [comment]
I've run into this issue a few times now. I'm going to have to do some investigation as to why this is happening and how to work around it.

**Comment by daryllabar** [comment]
I ran into this same error today...

**Comment by xairrick** [comment]
Happened to me today with a custom lookup on an activity.   Had to append the '_appointment' to the logicalname(yes it was the logicalname).   I looked that the OData Metadata and some Navigation Properties had appointment capitalize and  others didn't.   Damn MS for not be consistent.

**Comment by jlattimer** [comment]
Yeah as I've come to figure out the metadata is in fact not the source of truth - it's the CSDL.

Next update I'll have to change where I'm getting the data from. 

Found there are several little things like this all over.

**Comment by kip-dk** [comment]
If you name your custom lookup, using Capitalized letters ex. by writing

"Account" in the name when creating a reference attribute, , the UI will suggestion yourprefix_AccountId,  and if you keep that, you must 
use "yourprefix_AccountId" when updating references, while the property is returned as

_yourprefix_accountid_value in a fetch.

not a perfect api!. grrrr.

**Comment by anubhavbajpai** [comment]
Hi Jason,
Just a quick update, the organization which is using the webapi version 8.1 uses schemaname as the fieldname for the lookup attributes whereas webapi version8.2 uses "name" as the fieldname for the lookup attributes.

(Please note that the new code should be able to handle both the v8.1 as well as v8.2 environments instead of keeping two versions of code.)
Before making the changes for the new version, it would be better to query for the webapi version 8.2, if it is available then use the name fields for the lookup attributes. Otherwise if version 8.2 is not available and version 8.1 is available or lower then use schema names.

This is really frustating for such webapi changes. Atleast Microsoft should handle this in their underlying code.

I also tried to call the queries using v8.1 in the v8.2 environment then also its v8.1 api behaves like v8.2 only.
